src/Transfer.js

```diff
@@ -0,0 +1,390 @@
+import {
+    singleSelectedPropType,
+    multiSelectedPropType,
+} from './common-prop-types.js'
+
+/**
+ * @module
+ * @param {Transfer.PropTypes} props
+ *
+ * @returns {React.Component}
+ *
+ * @example import { Transfer } from @dhis2/ui-core
+ * @see Specification: {@link https://github.com/dhis2/design-system/blob/master/organisms/transfer.md|Design system}
+ * @see Live demo: {@link /demo/?path=/story/transfer--basic|Storybook}
+ */
+export const Transfer = ({
+    onChange,
+
+    children,
+    className,
+    dataTest,
+    disabled,
+    emptyOptionsPlaceholder,
+    emptySelectionPlaceholder,
+    enableOrderChange,
+    filter,
+    filterLabel,
+    filterOptions,
+    height,
+    initialFilter,
+    labelAddAll,
+    labelAddIndividual,
+    labelRemoveAll,
+    labelRemoveIndivitual,
+    leftFooter,
+    leftHeader,
+    maxSelections,
+    optionsWidth,
+    rightFooter,
+    selected,
+    selectedWidth,
+    showFilterInput,
+    onFilterChange,
+}) => {
+    /**
+     * As this component will filter and sort children
+     * in the selected area, adding keys will prevent
+     * reconciliation from suffering badly
+     */
+    const allOptions = addKeyToReactOptions(children)
+
+    /**
+     * Used in the "Filter" section and for
+     * limiting the selectable source options
+     *
+     * Filter can be controlled & uncontrolled.
+     * Providing the "onFilterChange" callback will make it a controlled value
+     */
+    const isFilterEnabled = showFilterInput || onFilterChange
+    const [internalFilter, setInternalFilter] = useState(initialFilter)
+    const actualFilter = onFilterChange ? filter : internalFilter
+
+    /**
+     * These are all the available option react elements.
+     * It will replace all selected options with null
+     */
+    const sourceOptions = filterOutOptions(allOptions, selected)
+    const filteredSourceOptions = getSubsetByFilter({
+        reactOptions: sourceOptions,
+        filter: actualFilter,
+        isFilterEnabled,
+        filterOptions,
+    })
+
+    /**
+     * Extract the selected options. This way custom options are supported
+     * without having to provide a component via the props
+     *
+     * Children are sorted by the order given in the "selected" array.
+     * This is done in order to cover the "append newly selected items
+     * at the end" feature/behavior.
+     */
+    const pickedOptions = extractPickedReactOptions({
+        reactOptions: allOptions,
+        selectedPlainOptions: selected,
+    })
+
+    /**
+     * These are all the highlighted option react elements on the options side.
+     */
+    const {
+        highlightedOptions: highlightedSourceOptions,
+        setHighlightedOptions: setHighlightedSourceOptions,
+        toggleHighlightedOption: toggleHighlightedSourceOption,
+    } = useHighlightedOptions({
+        availableReactOptions: filteredSourceOptions,
+        disabled,
+        maxSelections,
```

Comment:
There are a couple of option types, naming wise, in this component. We have:

* source options
* picked options
* selected options
* available options
* the filtered variants
* and the react vs plain object variants

I think that understanding those different option types is a big part to understanding this component. Naming wise, source, picked, filtered and the react vs plain all make sense to me.

* I'm not sure if there is a difference between selected and picked though?
* I see you're using available as well, I assume that that is for when a function takes options, but they're not always of the same type? I that case we could maybe just omit available?

1. I think that, after clearing up the above, a comment block at the top that clearly defines what the different option types signify would help devs a lot in understanding the component.
2. For the select we also deal with plain and react options. But there, we only call react options `reactOptions`, and the plain options just `options`. I think your distinction is clearer though, so I'm in favour of applying the same naming strategy for the select.

Reply:
Made a proposal here: https://dhis2.slack.com/archives/CBM8LNEQM/p1586440987039200

Reply:
"available" is not gone.

Reply:
Discussion over: We keep the current names & I've added a legend in the doc block of the component